pkg/controller/import-controller.go

```diff
@@ -23,70 +28,147 @@ const (
+		pvcLister:       pvcInformer.Lister(),
+		podLister:       podInformer.Lister(),
+		pvcsSynced:      pvcInformer.Informer().HasSynced,
+		podsSynced:      podInformer.Informer().HasSynced,
+		importerImage:   importerImage,
+		pullPolicy:      pullPolicy,
+		verbose:         verbose,
+		podExpectations: expectations.NewUIDTrackingControllerExpectations(expectations.NewControllerExpectations()),
 	}
 
 	// Bind the pvc SharedIndexInformer to the pvc queue
 	c.pvcInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
-		AddFunc: func(obj interface{}) {
-			key, err := cache.MetaNamespaceKeyFunc(obj)
-			if err == nil {
-				c.pvcQueue.AddRateLimited(key)
-			}
-		},
-		// this is triggered by an update or it will also be
-		// be triggered periodically even if no changes were made.
+		AddFunc: c.enqueuePVC,
 		UpdateFunc: func(old, new interface{}) {
-			key, err := cache.MetaNamespaceKeyFunc(new)
-			if err == nil {
-				c.pvcQueue.AddRateLimited(key)
-			}
+			c.enqueuePVC(new)
 		},
 	})
 
 	// Bind the pod SharedIndexInformer to the pod queue
 	c.podInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
-		AddFunc: func(obj interface{}) {
-			key, err := cache.MetaNamespaceKeyFunc(obj)
-			if err == nil {
-				c.podQueue.AddRateLimited(key)
-			}
-		},
-		UpdateFunc: func(oldObj, newObj interface{}) {
-			key, err := cache.MetaNamespaceKeyFunc(newObj)
-			if err == nil {
-				c.podQueue.AddRateLimited(key)
+		AddFunc: c.handlePodAdd,
+		UpdateFunc: func(old, new interface{}) {
+			newDepl := new.(*v1.Pod)
+			oldDepl := old.(*v1.Pod)
+			if newDepl.ResourceVersion == oldDepl.ResourceVersion {
+				// Periodic resync will send update events for all known PVCs.
+				// Two different versions of the same PVCs will always have different RVs.
+				return
 			}
+			c.handlePodUpdate(new)
 		},
+		DeleteFunc: c.handlePodDelete,
 	})
 
 	return c
 }
 
+func (c *ImportController) handlePodAdd(obj interface{}) {
+	c.handlePodObject(obj, "add")
+}
+func (c *ImportController) handlePodUpdate(obj interface{}) {
+	c.handlePodObject(obj, "update")
+}
+func (c *ImportController) handlePodDelete(obj interface{}) {
+	c.handlePodObject(obj, "delete")
+}
+
+func (c *ImportController) expectPodCreate(pvcKey string) {
+	c.podExpectations.ExpectCreations(pvcKey, 1)
+}
+func (c *ImportController) observePodCreate(pvcKey string) {
+	c.podExpectations.CreationObserved(pvcKey)
+}
+
+func (c *ImportController) handlePodObject(obj interface{}, verb string) {
+	var object metav1.Object
+	var ok bool
+	if object, ok = obj.(metav1.Object); !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			runtime.HandleError(errors.Errorf("error decoding object, invalid type"))
+			return
+		}
+		object, ok = tombstone.Obj.(metav1.Object)
+		if !ok {
+			runtime.HandleError(errors.Errorf("error decoding object tombstone, invalid type"))
+			return
+		}
+		glog.V(4).Infof("Recovered deleted object '%s' from tombstone", object.GetName())
```

Comment:
4 is not currently cdi verbosity level -- see similar comments in dataVolume controller.